src/coprocessor/dag/batch/executors/slow_hash_aggr_executor.rs

```diff
@@ -0,0 +1,401 @@
+            groups: HashMap::default(),
+            group_by_exps,
+            states_offset_each_row: Vec::with_capacity(
+                crate::coprocessor::dag::batch_handler::BATCH_MAX_SIZE,
+            ),
+        };
+
+        Ok(Self(AggregationExecutor::new(
+            summary_collector,
+            aggr_impl,
+            src,
+            config,
+            aggr_defs,
+            aggr_def_parser,
+        )?))
+    }
+}
+
+pub struct SlowHashAggregationImpl {
+    states: Vec<Box<dyn AggrFunctionState>>,
+    groups: HashMap<Vec<u8>, GroupInfo>,
+    group_by_exps: Vec<RpnExpression>,
+    states_offset_each_row: Vec<usize>,
+}
+
+#[derive(Debug)]
+struct GroupInfo {
+    /// The start offset of the states of this group stored in
+    /// `SlowHashAggregationImpl::states`.
+    states_start_offset: usize,
+
+    /// The offset of each GROUP BY column in the group key. There will be an additional offset
+    /// at the end, which equals to the group key length.
+    ///
+    /// For example, if the group key `AaaFfffff` is composed by two group column `[Aaa, Ffffff]`,
+    /// then the value of this field is `[0, 3, 9]`.
+    ///
+    /// Note 1:
+    /// We want aggregation executor to provide each GROUP BY column independently, which will be
+    /// useful in future executors (like Projection executor). However, currently this segmentation
+    /// is not useful because we return data by row and columns will be always pasted together.
+    /// Thus segmentation or not doesn't make any differences.
+    ///
+    /// Note 2:
+    /// We use `SmallVec<[..; 6]>` so that when group by columns <= 5 SmallVec avoids allocation.
+    /// We think the case that there are more than 5 group by columns are rare.
+    ///
+    /// Note 3:
+    /// We use `SmallVec<[u32; ..]>` instead of `SmallVec<[usize; ..]>` because the length of the
+    /// group key will not exceed `u32`. Using `u32` instead of `usize` makes this field small
+    /// and can be fit into the cache line better.
+    group_key_offsets: SmallVec<[u32; 6]>,
+}
+
+impl<Src: BatchExecutor> AggregationExecutorImpl<Src> for SlowHashAggregationImpl {
+    #[inline]
+    fn prepare_entities(&mut self, entities: &mut Entities<Src>) {
+        let src_schema = entities.src.schema();
+        for group_by_exp in &self.group_by_exps {
+            entities
+                .schema
+                .push(group_by_exp.ret_field_type(src_schema).clone());
+        }
+    }
+
+    #[inline]
+    fn process_batch_input(
+        &mut self,
+        entities: &mut Entities<Src>,
+        mut input: LazyBatchColumnVec,
+    ) -> Result<()> {
+        let rows_len = input.rows_len();
+
+        // 1. Calculate which group each src row belongs to.
+        self.states_offset_each_row.clear();
+
+        let src_schema = entities.src.schema();
+
+        // TODO: Eliminate these allocations.
+        let mut group_by_keys = Vec::with_capacity(rows_len);
```

Comment:
Can we move the `Vec`s to the struct to avoid creating repeatedly? I think it's relatively easy but should help.

Reply:
Yes we can, but it will only save one allocation. As a comparison, we already allocate 1024 vectors here 😂 So I decided to not to do this to make it looks simple.